### PR TITLE
Added support for direct initialization

### PR DESCRIPTION
**Pull request by apollo13**
This is a re-implementation of #14 without any backwards incompatible changes (aside from a pydantic version increase)

### CODE REVIEW

File: goodconf/__init__.py [hunk 1]
```diff
@@ -65,26 +67,57 @@ def initial_for_field(name: str, field: ModelField) -> Any:
...
-        if load:
-            self.load()
+        self._config_file = None
+        # Emulate Pydantic behavior, load immediately
+        if kwargs:
+            return super().__init__(**kwargs)
+        elif load:
+            return self.load()
+
+    _config_file: str | None = PrivateAttr()
```
Comment (apollo13) [inline goodconf/__init__.py:111]: This requires Pydantic 1.7, if this is not okay for you then I can work around this.

File: tests/test_goodconf.py [hunk 2]
```diff
@@ -154,3 +157,46 @@ class TestConf(GoodConf):
...
+        c = TestConf(load=True)
+
+    assert c.a
+
+
+def test_precedence(tmpdir):
+    path = tmpdir.join("myapp.json")
+    path.write(json.dumps({"init": "file", "env": "file", "file": "file"}))
+
+    class TestConf(GoodConf, default_files=[path]):
```
Comment (apollo13) [inline tests/test_goodconf.py:188]: Did you know that you do not need to use `class Config`? Pydantic allows you to pass those via the metaclass kwargs.

File: tests/test_goodconf.py [hunk 3]
```diff
@@ -154,3 +157,46 @@ class TestConf(GoodConf):
...
+    class TestConf(GoodConf):
+        a: bool = False
+
+        class Config:
+            env_prefix = "PREFIX_"
+
+    with env_var("PREFIX_A", "True"):
+        c = TestConf(load=True)
+
+    assert c.a
```
Comment (apollo13) [inline tests/test_goodconf.py:181]: This shows that #10 can be closed, it is supported…

File: tests/test_goodconf.py (hunk 2 above)
Comment (ipmb) [reply tests/test_goodconf.py:188]: I did not!

File: pyproject.toml [hunk 4]
```diff
@@ -21,7 +21,7 @@ classifiers = [
 ]
 dynamic = ["version"]
 dependencies = [
-    "pydantic>=1.0"
+    "pydantic>=1.7,<2.0"
```
Comment (ipmb) [inline pyproject.toml:24]: Are we sure 2.0 will be breaking for our use case? I'm hesitant to add the upper version restriction.

File: goodconf/__init__.py [hunk 5]
```diff
@@ -94,34 +127,21 @@ def customise_sources(
...
             """Load environment variables before init"""
-            return env_settings, init_settings, file_secret_settings
-
-    def load(self, filename: str = None) -> None:
+            return (
+                init_settings,
+                env_settings,
+                file_config_settings_source,
+                file_secret_settings,
+            )
```
Comment (ipmb) [inline goodconf/__init__.py:135]: Could you document the precedence here in the README?

**Comment by apollo13** [comment]
Will document the precedence. But just so there is no misunderstanding, for all intents and purposes the precedence stays as it was before. That is ENV wins over file.cfg and initialization wins over everything else (which wasn't possible before).

As for the pydantic version boundary: will remove it, we can always issue a new release if we find out that 2.0 is not compatible.

File: goodconf/__init__.py (hunk 5 above)
Comment (apollo13) [reply goodconf/__init__.py:135]: I just looked at the README and it already documents the precedence of env over files. I'd prefer not to document anything else currently and wait to see if pydantic 2 would require further changes. In the same sense I would not document initialization via `__init__` now. Everyone using this library and knows pydantic can use it, but let's wait and see what pydantic 2 does before making this an official feature.

**Comment by ipmb** [comment]
@apollo13 could you update `CHANGES.rst` and I can cut a new release? We should probably call this `2.1.0`

**Pull request merged**